fuzz_test.go

```diff
@@ -0,0 +1,103 @@
+package smt
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/binary"
+	"math"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func FuzzSMT(f *testing.F) {
+	seeds := [][]byte{
+		[]byte(""),
+		[]byte("foo"),
+		{1, 2, 3, 4},
+		[]byte("\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"),
+		nil,
+	}
+	for _, s := range seeds {
+		f.Add(s)
+	}
+	f.Fuzz(func(t *testing.T, input []byte) {
+		t.Log(input)
+		smn := NewSimpleMap()
+		tree := NewSparseMerkleTree(smn, sha256.New())
+
+		r := bytes.NewReader(input)
+		var keys [][]byte
+		key := func() []byte {
+			if readByte(r) < math.MaxUint8/2 {
+				k := make([]byte, readByte(r)/2)
+				if _, err := r.Read(k); err != nil {
+					return nil
+				}
+				keys = append(keys, k)
+				return k
+			}
+
+			if len(keys) == 0 {
+				return nil
+			}
+
+			return keys[int(readByte(r))%len(keys)]
+		}
+
+		for i := 0; r.Len() != 0; i++ {
```

Comment:
What you want is go's equivalent of a `while loop` rather than a `for loop`.

```
for r.Len() > 0 {
}
```

Reply:
So we need the `i` counter as when we do the update operation that is the value we pass - I am not checking the root in each op and its more efficient to use the counter than randomly generate a variable - incase two keys are the same we want to check the root does change.